### ISSUE TITLE
Huge issue with time-only DateTimeField and DST

### ISSUE THREAD

**Issue by dpomt** [OrchardCMS/Orchard#6731]
Hi, 

I have a huge issue with time-only DateTimeField in a productive site running Orchard 1.9.x. 

The issue is that time-only fields are always stored as UTC (converted from CurrenetTimeZone at moment of saving/publishing a content item) time but rendered as local time (at the moment of displaying the content item) and this depends if DST is active for moment of creation and/or displaying the field. 

**First for clarification:**
Central European Summer Time (CEST) = Central European Daylight Saving Time (CEDT= CET DST). CEDT = UTC + 2. 
Central Summer Time (CEST) = Central European Daylight Saving Time (CEDT= CET DST). CET = UTC + 1. 

**Example:**
At my Event content type, I have one date-only field for the date and one time-only field for the time. I took this approach since it should be possible for an event to have no time. 
Regard the following 4 cases all dealing with creation of a time for a CET date - they differ only in creation date (whether CET or CEDT) and display date (whether CET or CEDT) 

**Case 1:** On November 2015 (CET) create an event with a date for May 2016 (CEDT). Set the time-only field to 10:00 a.m. Since creation is while CET it is stored as 09:00 a.m. (UTC=CET-1). 
**Case 1a:** Look at the item while CET, e. g. January 2016: time is displayed correctly, since CET = UTC+1 ==> 10:00 a.m. 
**Case 1b:** Look at the item while CEDT, e. g. May 2016: time is displayed incorrectly, since CEDT = UTC+2 ==> 11:00 a.m. 

**Case 2:** On April 2016 (CEDT) create an event with a date for May 2016 (CEDT). Set the time-only field to 10:00 a.m. Since creation is while CEDT it is stored as 08:00 a.m. (UTC=CEDT-2). 
**Case 2a:** Look at the item while CEDT, e. g. May 2016: time is displayed correctly, since CEDT = UTC+2 ==> 10:00 a.m. 
**Case 2b:** Look at the item while CET, e. g. January 2017: time is displayed incorrectly, since CET = UTC+1 ==> 09:00 a.m. 

I took a look at the implementation in DefaultDateLocalizationService:ConvertFromLocalizedString and DateTimeFieldDriver:Display/Editor. 
I think to fix the issue we need to add "options.EnableTimeZoneConversion = false;" for time-only fields in the DateTimeFieldDriver:Display and in the DateTimeFieldDriver:Editor methods. 

Totally there are two occurences where I applied added here disabling the time zone conversion: 
if (settings.Display == DateTimeFieldDisplays.TimeOnly) {
                options.EnableCalendarConversion = false;
                options.EnableTimeZoneConversion = false;
                options.IgnoreDate = true
} 

Of course with this fix all existing time-only field values stored in DB are also incorrect and need to be adapted.

Please give me feedback if this is a suitable fix or if there is any better way to fix it. 

Thanks, 
Dieter

**Comment by DanielStolt** [comment]
It's complicated, took several reads, but I think maybe I understand the issue now. :)

A time-only field has no concept of a date, so how do we decide if the time is DST or not? I guess the approach I took when implementing this was to base it on the current date, at the time of storage/retrieval.

I don't agree that the correct solution is to remove *time zone conversion*. We should, however, remove *DST conversion* since the concept does not even apply if there is no date.

I'm wondering if perhaps one solution would be to simply make it deterministic, i.e. when storing/retrieving a time-only value, always base the conversion on a fixed date, such as `DateTime.MinValue.Date`. This would ensure that, whenever you save the field and whenever you render it, the result would always be the same with respect to DST, and the time zone conversion would still work as expected.

What do you think of that @dpomt?

**Comment by sebastienros** [comment]
Interestingly enough, I was just looking at this field for an unrelated issue, and discovered that it was storing different information based on the settings (Date only) which I think is wrong. DateTime should only store a date time, and display the time only if required, which would solve this issue too. Or if we care about only a Time that is invariant, then a new field is necessary. So I think we should update the field to store a full date and time. The migrations would be transparent as today's date would be taken into account during next load/save.

**Comment by DanielStolt** [comment]
@sebastienros I see problems with your suggestion. Which date do we store? Next time you save the content item, a different date would be stored, and then depending on DST differences between different time zones you would still see inconsistencies.

Do you see a problem with my suggestion?

**Comment by sebastienros** [comment]
I see the issue with my suggestion, yes I agree with your comment, but it still means to read the settings in the field then.
I really think the ultimate solution should be to have a DateTime field and a Time field (which would be time zone insensitive, just two ints)

Let's do with your suggestion though, this is the best non-breaking change we can do right now.

**Comment by dpomt** [comment]
@DanielStolt  I agree using a fixed date would be a solution and is better than removing time Zone conversion in general.

@sebastienros  I also agree that it would be even better to have a separate time field.
For my special usage having two DateTime fields for an event begin date/time, another Option would be to have a Checkbox in DateTime Editor to allow the user to flag it as an all day event. Right now, it is not possible to not specifiy the time portion within DateTime picker. If you leave it empty now it is stored as 00:00:00 UTC and then gets converted to e. g. UTC+2 when re-displayed in CEDT.


I am still hoping that we get rid of DST here in Germany ;) There are discussions on that.

**Comment by DanielStolt** [comment]
@sebastienros and @dpomt OK, assigning this to myself for implementation of the fixed date fix. As for the new separate "static time of day" field, that's better targeted at `dev` and perhaps should get its own separate issue to track.

**Comment by sebastienros** [comment]
https://github.com/OrchardCMS/Orchard/issues/6781

**Comment by dpomt** [comment]
Did you think about providing a migration step for fixing probably incorrect values? 

I did not find a suitable way to do this. For my productive site I corrected the value depending if a content item what created in DST or not. But there are IMO more cases than that (depending whether a content item was just created or duplicated, changed later or even if it was created prior to 1.8.x etc.). I ended up in sending mails to all persons that created events to check and adapt their event times.
